### PR TITLE
feat: add featuresShapCol to LightGBMClassifierModel

### PR DESCRIPTION
**Pull request by ocworld**
In this pr, featuresShapCol is added to LightGBMClassifierModel.

### CODE REVIEW

**Comment by ocworld** [comment]
@AhnLab-OSS

File: src/main/scala/com/microsoft/ml/spark/lightgbm/LightGBMClassifier.scala [hunk 1]
```diff
@@ -155,6 +160,11 @@ class LightGBMClassificationModel(
       outputData = outputData.withColumn(getLeafPredictionCol,  predLeafUDF(col(getFeaturesCol)))
       numColsOutput += 1
     }
+    if (getFeaturesShapCol.nonEmpty) {
+      val featureShapUDF = udf(featuresShap _)
+      outputData = outputData.withColumn(getLeafPredictionCol,  featureShapUDF(col(getFeaturesCol)))
```
Comment (imatiach-msft) [inline src/main/scala/com/microsoft/ml/spark/lightgbm/LightGBMClassifier.scala:165]: should this be: getLeafPredictionCol -> getFeaturesShapCol (?)

**Comment by ekerazha** [comment]
So we have SHAP for LightGBM Ranker and Classifier? Is it now missing only for Regressor?

File: src/main/scala/com/microsoft/ml/spark/lightgbm/LightGBMClassifier.scala [hunk 2]
```diff
@@ -86,16 +86,27 @@ trait HasLeafPredictionCol extends Params {
   def setLeafPredictionCol(value: String): this.type = set(leafPredictionCol, value)
 }
 
+trait HasFeaturesShapCol extends Params {
+  val featuresShapCol = new Param[String](this, "featuresShapCol",
+    "features' shap value column name that means their contribution for prediction")
```
Comment (imatiach-msft) [inline src/main/scala/com/microsoft/ml/spark/lightgbm/LightGBMClassifier.scala:91]: Maybe change the description to:
```
Output SHAP vector column name after prediction containing the feature contribution values
```

File: src/main/scala/com/microsoft/ml/spark/lightgbm/LightGBMClassifier.scala (hunk 1 above)
Comment (ocworld) [reply src/main/scala/com/microsoft/ml/spark/lightgbm/LightGBMClassifier.scala]: @imatiach-msft You're right. It is fixed now.

File: src/main/scala/com/microsoft/ml/spark/lightgbm/LightGBMClassifier.scala (hunk 2 above)
Comment (ocworld) [reply src/main/scala/com/microsoft/ml/spark/lightgbm/LightGBMClassifier.scala]: @imatiach-msft It is changed to your suggestion

**Comment by ocworld** [comment]
@ekerazha I assume that except regression SHAP is supported.
Please note that it has different usages for getting SHAP between LightGBM Ranker and Classifier in this pr.

**Comment by imatiach-msft** [comment]
hmm seems the test is failing in the native code:

```
[info] + Test Verify LightGBM Classifier features shap took 0.583s 
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f4ea7779ef2, pid=7254, tid=0x00007f4ea4263700
#
# JRE version: OpenJDK Runtime Environment () (8.0_252-b14) (build 1.8.0_252-b14)
# Java VM: OpenJDK 64-Bit Server VM (25.252-b14 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# C  [libc.so.6+0x7fef2]
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /home/vsts/work/1/s/hs_err_pid7254.log
[thread 139974389262080 also had an error]
#
# If you would like to submit a bug report, please visit:
#   http://www.azulsystems.com/support/
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
timeout: the monitored command dumped core
/home/vsts/work/_temp/azureclitaskscript1589903456723.sh: line 6:  7253 Aborted
```

**Comment by imatiach-msft** [comment]
running another build again, will try to run this branch locally also

**Comment by imatiach-msft** [comment]
@ocworld great news, I was able to run your code locally and I discovered the issue - SHAP returns the expected value in addition to the feature contributions, which led to memory corruption.  Although I was not able to get the SHAP test to fail, I did see another test fail due to memory corruption from the SHAP test, and once I fixed the issue I validated that the memory corruption error is no longer occurring.  I pushed the fix directly to your branch, hope you don't mind.

**Comment by imatiach-msft** [comment]
missed updating another test in ranker which was checking the number of features, fixed

**Comment by imatiach-msft** [comment]
found another minor issue where the new column name was missing in the objects to save which caused a serialization failure, fixed, hopefully build should be green now

**Comment by imatiach-msft** [comment]
once PR is merged I'll send a corresponding PR for
1.) pyspark API
2.) lightgbm regressor
to make it consistent across codebase

**Comment by imatiach-msft** [comment]
@ocworld I noticed this PR won't work for multiclass case, since multiclass shap values are of shape (# features + 1 expected_value) * # classes, but we currently aren't making the array the size of # classes.  I can fix this too when I send out the regression PR.

**Comment by ocworld** [comment]
@imatiach-msft Thanks for your notice

**Pull request merged**